z3/src/ast.rs

```diff
@@ -236,6 +236,39 @@ pub trait Ast<'ctx>: Sized + fmt::Debug {
             )
         })
     }
+
+    /// Return the number of children of this AST node.
+    /// Leaf nodes (eg Bool consts) will return 0.
+    fn num_children(&self) -> usize {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let this_ctx = self.get_ctx().z3_ctx;
+            let this_app = Z3_to_app(this_ctx, self.get_z3_ast());
+            Z3_get_app_num_args(this_ctx, this_app) as usize
+        }
+    }
+
+    /// Return the nth child of this AST node.  Out-of-bound indices will panic.
+    fn nth_child(&self, idx: usize) -> Dynamic<'ctx> {
+        if idx >= self.num_children() {
+            panic!("Index out of bounds: {} max index {}", idx, self.num_children());
+        }
```

Comment:
Rather than panicking, let's change the function's signature to returning `Option<Dynamic<'ctx>>` and return `None` here.